org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semantictokens/SemanticTokensService.java

```diff
@@ -0,0 +1,224 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.common.util.WrappedException;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.SemanticTokens;
+import org.eclipse.lsp4j.SemanticTokensParams;
+import org.eclipse.lsp4j.SemanticTokensRangeParams;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.HighlightingStyles;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.ISemanticHighlightingCalculator;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.LightweightPosition;
+import org.eclipse.xtext.ide.editor.syntaxcoloring.MergingHighlightedPositionAcceptor;
+import org.eclipse.xtext.ide.server.Document;
+import org.eclipse.xtext.ide.server.UriExtensions;
+import org.eclipse.xtext.resource.IResourceServiceProvider;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.inject.Inject;
+
+/**
+ * The Class SemanticTokensService.
+ */
+public class SemanticTokensService {
+	@Inject
+	private UriExtensions uriExtensions;
+	@Inject
+	private IResourceServiceProvider.Registry languagesRegistry;
+	private List<String> tokenTypes = new ArrayList<>();
+
+	SemanticTokensService() {
+		addTokenTypes();
+	}
+
+	protected void addTokenTypes() {
+		tokenTypes.add(HighlightingStyles.KEYWORD_ID);
+		tokenTypes.add(HighlightingStyles.DEFAULT_ID);
+		tokenTypes.add(HighlightingStyles.NUMBER_ID);
+	}
+
+	private List<LightweightPosition> getPositions(final XtextResource resource,
+			final TextDocumentIdentifier textDocumentIdentifier, final CancelIndicator cancelIndicator) {
+		URI uri = getURI(textDocumentIdentifier);
+		ISemanticHighlightingCalculator semanticHighlightingCalculator = getService(uri,
+				ISemanticHighlightingCalculator.class);
+		if (semanticHighlightingCalculator != null) {
+		MergingHighlightedPositionAcceptor acceptor = new MergingHighlightedPositionAcceptor(
+				semanticHighlightingCalculator);
+	    	semanticHighlightingCalculator.provideHighlightingFor(resource, acceptor, cancelIndicator);
+	    	acceptor.mergePositions();
+		return acceptor.getPositions();
+		}
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Semantic tokens full.
+	 *
+	 * @param params
+	 *            the params
+	 * @param cancelIndicator
+	 *            the cancel indicator
+	 * @return the semantic tokens
+	 */
+	public SemanticTokens semanticTokensFull(final Document document, final XtextResource resource,
+			final SemanticTokensParams params, final CancelIndicator cancelIndicator) {
+	    if (document == null || resource == null) {
+	        return new SemanticTokens(Collections.emptyList());
+	    }
+	    List<Integer> data = new ArrayList<>();
+		int lastOffset = 0;
+		int lastLine = 0;
+	    try (PositionReader reader = new PositionReader(document.getContents())) {
```

Comment:
Instead of using a reader on the document, did you try to use `org.eclipse.xtext.nodemodel.util.NodeModelUtils.getLineAndColumn(INode, int)` to compute the positions?

If necessary, we can probably expose `org.eclipse.xtext.nodemodel.impl.InternalNodeModelUtils.getLineBreakOffsets(INode)` from `NodeModelUtils` and use the ordered positions to do a pairwise walk over the offsets and the positions.